src/Command/Ssh/SshKeyCommandBase.php

```diff
@@ -31,15 +66,131 @@ public static function getIdeSshKeyLabel(IdeResponse $ide): string {
+  }
 
-    return $label;
+  /**
+   * Normalizes public SSH key by trimming and removing user and machine suffix.
+   *
+   * @param string $public_key
+   *
+   * @return string
+   */
+  protected function normalizePublicSshKey($public_key): string {
+    $parts = explode('== ', $public_key);
+    $key = $parts[0];
+
+    return trim($key);
+  }
+
+  /**
+   * Asserts whether ANY SSH key has been added to the local keychain.
+   *
+   * @return bool
+   * @throws \Exception
+   */
+
+  protected function sshKeyIsAddedToKeychain(): bool {
+    $process = $this->localMachineHelper->execute([
+      'ssh-add',
+      '-L',
+    ], NULL, NULL, FALSE);
+
+    if ($process->isSuccessful()) {
+      $key_contents = $this->normalizePublicSshKey($this->localMachineHelper->readFile($this->publicSshKeyFilepath));
+      return strpos($process->getOutput(), $key_contents) !== FALSE;
+    }
+    return FALSE;
+  }
+
+  /**
+   * Adds a given password protected local SSH key to the local keychain.
+   *
+   * @param string $filepath
+   *   The filepath of the private SSH key.
+   * @param string $password
+   *
+   * @throws \Acquia\Cli\Exception\AcquiaCliException
+   */
+  protected function addSshKeyToAgent(string $filepath, string $password): void {
+    // We must use a separate script to mimic user input due to the limitations of the `ssh-add` command.
+    // @see https://www.linux.com/topic/networking/manage-ssh-key-file-passphrase/
+    $temp_filepath = $this->localMachineHelper->getFilesystem()->tempnam(sys_get_temp_dir(), 'acli');
+    $this->localMachineHelper->writeFile($temp_filepath, <<<'EOT'
+#!/usr/bin/env bash
+echo $SSH_PASS
+EOT
+    );
+    $this->localMachineHelper->getFilesystem()->chmod($temp_filepath, 0755);
+    $private_key_filepath = str_replace('.pub', '', $filepath);
+    $process = $this->localMachineHelper->executeFromCmd('SSH_PASS=' . $password . ' DISPLAY=1 SSH_ASKPASS=' . $temp_filepath . ' ssh-add ' . $private_key_filepath, NULL, NULL, FALSE);
+    $this->localMachineHelper->getFilesystem()->remove($temp_filepath);
+    if (!$process->isSuccessful()) {
+      throw new AcquiaCliException('Unable to add the SSH key to local SSH agent:' . $process->getOutput() . $process->getErrorOutput());
+    }
+  }
+
+  /**
+   * Polls the Cloud Platform until a successful SSH request is made to the dev
+   * environment.
+   *
+   * @param \Symfony\Component\Console\Output\OutputInterface $output
+   *
+   * @throws \Exception
+   */
+  protected function pollAcquiaCloudUntilSshSuccess(
+    OutputInterface $output
+  ): void {
+    // Create a loop to periodically poll the Cloud Platform.
+    $loop = Loop::get();
+    $spinner = LoopHelper::addSpinnerToLoop($loop, 'Waiting for the key to become available on the Cloud Platform', $output);
+
+    // Wait for SSH key to be available on a web.
+    $cloud_app_uuid = $this->determineCloudApplication(TRUE);
+    $environment = $this->getAnyAhEnvironment($cloud_app_uuid);
```

Comment:
Sorry, maybe I should have been more specific before, because the same problem of checking "any" application applies to checking "any" environment for a given app. Many users will have SSH access to dev/stage, but not prod. This check will thus time out for many developers.

Let's modify `getAnyAhEnvironment()` to prefer dev/stage. It's a heuristic check but seems safe, I can't imagine many people have access to prod but _not_ dev/stage. The alternative is asking people to select a specific environment (confusing UX), or polling the permissions API to find an environment with SSH access (complex).